### PR TITLE
Constraints modal re design

### PR DESCRIPTION
**Pull request by rajatvijay**
Questions: 
- [x] There are no design/help texts for the `check` and `exclude` constraint types. Is this intentional?
> Yes, we don't support showing / creating those constraints currently because you can have arbitrary expressions in the constraint definition. We'll probably support these when we support formulas in the product.
- [x] When creating a new constraint the option to choose the name auto/manual is missing. Is this intentional?
> Yes, there's no additional toggle. We show the constraint name in an input which should be auto-filled with the automatic name. Users can edit it if they want to change the name.
- [x] There are designs for `null` constraints. Is this something new?
> https://github.com/centerofci/mathesar/issues/2114

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by kgodey** [comment]
> There are no design/help texts for the `check` and `exclude` constraint types. Is this intentional?

Yes, we don't support showing / creating those constraints currently because you can have arbitrary expressions in the constraint definition. We'll probably support these when we support formulas in the product.

> When creating a new constraint the option to choose the name auto/manual is missing. Is this intentional?

Yes, there's no additional toggle. We show the constraint name in an input which should be auto-filled with the automatic name. Users can edit it if they want to change the name.

> There are designs for `null` constraints. Is this something new?

It should actually be `NOT NULL` constraints, sorry I didn't notice that in the designs. It is new, but it's also the same thing as the restrict `NULL` checkbox in the columns.

>  - Unique constraint error toast.
>  - Edit tooltip for the constraint name.

Not sure what the questions are here.

**Review by seancolsen — CHANGES_REQUESTED** [review]
Hey @rajatvijay, a few small changes to request.

- The trash button is missing here

    | Figma | App |
    | -- | -- |
    | image | image |

- When I add a new constraint, the modal closes. It would be nice if we could keep it open. If that's too hard, then we can open a ticket and do it later.

- After I delete a PK constraint, I don't seem to be able to add a new one again. I think we should support that. If you don't want to handle it in this PR, then you can open a ticket.

- I added some tickets for things we can address later

    - #2101
    - #2102

File: mathesar_ui/src/component-library/multi-select/MultiSelect.scss [hunk 1]
```diff
@@ -21,11 +21,10 @@
 
   .selected-value {
     display: flex;
-    padding: 0 var(--spacing-2);
+    padding: 0.285rem 0.428rem;
     margin: var(--spacing);
-    background: #f4f4f4;
```
[the remark this answers is not in the document]
Comment (ghislaineguerin) [reply mathesar_ui/src/component-library/multi-select/MultiSelect.scss:26]: @seancolsen that was not intentional. I do agree that the background one looks better and we shouldn't change it.

File: mathesar_ui/src/component-library/labeled-input/LabeledInput.scss [hunk 2]
```diff
@@ -5,6 +5,7 @@
   .label {
     display: inline-block;
     color: var(--slate-800);
+    font-size: var(--text-size-small);
```
[the remark this answers is not in the document]
Comment (ghislaineguerin) [reply mathesar_ui/src/component-library/labeled-input/LabeledInput.scss:8]: @seancolsen Agreed, all labels should have the same size. If any specific style needs to be overwritten I'll handle that during the consistency work. thanks.

**Comment by rajatvijay** [comment]
> > When creating a new constraint the option to choose the name auto/manual is missing. Is this intentional?
> 
> Yes, there's no additional toggle. We show the constraint name in an input which should be auto-filled with the automatic name. Users can edit it if they want to change the name.

17a56b65

**Comment by rajatvijay** [comment]
> * The trash button is missing here

Confirmed with @ghislaineguerin, this is not needed here. 

> When I add a new constraint, the modal closes. It would be nice if we could keep it open. If that's too hard, then we can open a ticket and do it later.
Seems to be happening randomly. (Sorry for the blurry video, I'll have to adjust some settings on the loom app to fix this. But this will demo the point well enough) https://www.loom.com/share/b33ddceb4f0648a99c1aebc1c09849f4
Here is the issue for this: https://github.com/centerofci/mathesar/issues/2115

> After I delete a PK constraint, I don't seem to be able to add a new one again. I think we should support that. If you don't want to handle it in this PR, then you can open a ticket.
According to the designs we should not support deleting primary key constraints, hence: 17d1d4a7. 
@kgodey do we need a ticket to handle this in the future?

File: mathesar_ui/src/systems/table-view/constraints/ConstraintTypeSection.svelte [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  import ConstraintCollapseHeader from './ConstraintCollapseHeader.svelte';
+  import NewUniqueConstraint from './NewUniqueConstraint.svelte';
+  import NewFkConstraint from './NewFkConstraint.svelte';
+  import ConstraintDetails from './ConstraintDetails.svelte';
+  import ForeignKeyConstraintDetails from './ForeignKeyConstraintDetails.svelte';
+
+  export let constraintType: ConstraintType;
+  export let constraints: Constraint[];
+
+  let addConstraintType: 'foreignKey' | 'unique' | 'none' = 'none';
```
[the remark this answers is not in the document]
Comment (rajatvijay) [reply mathesar_ui/src/systems/table-view/constraints/ConstraintTypeSection.svelte]: 5dc5f014

File: mathesar_ui/src/component-library/labeled-input/LabeledInput.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (rajatvijay) [reply mathesar_ui/src/component-library/labeled-input/LabeledInput.scss]: c03141a2

File: mathesar_ui/src/component-library/multi-select/MultiSelect.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (rajatvijay) [reply mathesar_ui/src/component-library/multi-select/MultiSelect.scss:26]: c03141a2

**Review by seancolsen — CHANGES_REQUESTED** [review]
@rajatvijay you have addressed all of my critique.

_However_, I want to discuss the changes you made in 17a56b65fce18a6a92842ffc4aa0599ea630fd2d to address Kriti's critique.

@kgodey and @ghislaineguerin I'm looping you into this conversation because I have some important concerns with the design that I think we need to address before merging this PR.

Let's see what 17a56b65fce18a6a92842ffc4aa0599ea630fd2d did:

| Before | After |
| -- | -- |
| image | image |

There are two problems with this change:

- (1) _(an implementation problem)_ The constraint name does not automatically update to reflect the columns selected.

- (2) _(a design problem)_ The front end has no way to distinguish between the following two scenarios:

    - (A) The user cares how the constraint is named, and the name Mathesar suggested happens to agree with the user's desired name. Here, the front end needs send the constraint name to the server.

    - (B) The user doesn't care how the constraint is named. (This is scenario is much more likely.) Here we should _not_ send the constraint name to the server.

    Because the front end has no way to distinguish between (A) and (B), it must always send the constraint name to the server. Responsibility is thus delegated entirely to the front end to suggest constraint names. This change in behavior is rather problematic because the server is much better at auto-generating constraint names since it has access to more data. The constraint name needs to be unique across all the constraint names in all tables _and_ all the table names in the schema. The front end does not have all that data so it can't suggest a name that it knows to be unique. Sure, it can suggest names prefixed by the table name. But if tables are renamed later, the constraints are not, so there's still opportunity for problems here. I've argued many times that the front end _should_ have the full schema structure in memory. That architecture would allow us to much more safely implement the new design here. But with our current architecture, I think we ought to make an effort to push this constraint-naming logic to the back end as much as we can.

    Further, from a UX perspective, I'd argue that the previous design is superior because it allows the user to do less thinking while filling out the form. The user only has two options to consider at first and the help bubble explains the difference between the options (while strongly suggesting the default option). Without the radio choices, the user is left to wonder what implications the constraint name will have and whether they _should_ customize it. I feel pretty strongly that we should steer users away from customizing constraint names. If they happen to name a constraint "Foo", then it becomes impossible to name a new table "Foo" -- and that limitation is entirely obscured by the new design.

**I suggest reverting 17a56b65fce18a6a92842ffc4aa0599ea630fd2d, keeping the old design.**

**If we do continue with the new design, then:**

- We need to fix the implementation problem described above, otherwise adding two constraints to the same table will fail unless the user manually assigns names.

    We should make sure that the suggested constraint name is unique across all table names and all constraint names within the current table (because the front end _does_ have that data). And we should suffix the suggested name with `_key` or similar, just to guard against the likelihood that a user will want that name for a table later on.

- I think we should add some help text like we had before.

    image

**Comment by kgodey** [comment]
Okay, I'm fine with keeping the old design and adding help text. The Figma design was confusing because it showed both the radio toggle AND the constraint name input all the time, so I suggested getting rid of the toggle.

**Comment by ghislaineguerin** [comment]
I agree with keeping the old design.

**Review by seancolsen — APPROVED** [review]
Ok, I've reverted 17a56b65fce18a6a92842ffc4aa0599ea630fd2d. Merging this now. Thanks for your input, @kgodey and @ghislaineguerin

**Pull request merged**